### PR TITLE
Typing system overhaul. Improve support for object detection scenarios.

### PR DESCRIPTION
**Pull request by lrzpellegrini**
This PR contains a rewrite of many parts of the typing system of the benchmarks module. When analyzed using MyPy, the code in `utils`, `scenarios`, and `generators` folders no longer contain typing errors. Also, PyLance complains a lot less.

In addition, the elements needed to support object detection scenarios have been updated and separated from their classification counterparts.

This PR is currently a draft, but the most significant changes are already there. The next step is to extend this work to the remaining folders in the benchmarks module.

This is part of the ongoing work needed to support distributed training settings (#1315).

### CODE REVIEW

File: avalanche/benchmarks/utils/data.py [hunk 1]
```diff
@@ -237,22 +265,54 @@ def transform(self):
             "methods such as `replace_current_transform_group`. "
             "See the documentation for more info."
         )
+    
+    def _update_data_attribute(self, name: str, new_value):
+        """
+        Adds or replace a data attribute.
+
+        If a object of type :class:`DataAttribute` is passed, then the object is setted as is.
```
[the remark this answers is not in the document]
Comment (AntonioCarta) [reply avalanche/benchmarks/utils/data.py:274]: this should also be thoroughly tested

File: avalanche/benchmarks/scenarios/generic_scenario.py [hunk 2]
```diff
@@ -120,8 +336,49 @@ def __getattribute__(self, item):
...
+            v = super().__getattribute__(field)
+            if isinstance(v, ExperienceAttribute):
+                if v.use_in_train != use_in_train:
+                    raise RuntimeError(
+                        f'Experience attribute {field} redefined with inconguent use_in_train field. Was {v.use_in_train}, overridden with {use_in_train}.'
+                    )
+                
+                if v.use_in_eval != use_in_eval:
+                    raise RuntimeError(
+                        f'Experience attribute {field} redefined with inconguent use_in_eval field. Was {v.use_in_eval}, overridden with {use_in_train}.'
```
Comment (AntonioCarta) [inline avalanche/benchmarks/scenarios/generic_scenario.py:372]: typo

**Comment by AntonioCarta** [comment]
Ok, so I took a look at the PR. First of all, I believe that having types would users and reduce the complexity a little bit. However, in order to make them useful, we have to make some design changes. Right now, some design choices result in very complex types. IMO the worse culprit are the circular dependencies that we have between experiences->streams->benchmark.

I think we can remove them (i.e. not have `experience.origin_stream` and `stream.origin_benchmark`). This would require passing the benchmark object to the evaluator, which I think is a reasonable thing to do. The strategy should never touch the stream or benchmark object directly so it would be a benefit in this sense because it would remove the possibility of "cheating".

@lrzpellegrini what do you think? We don't have to do the change in this PR (actually, I would prefer a lot if we didn't, it's already huge), but we should make a choice.

If we keep the the current design it's going to be very difficult to add new benchmarks without breaking the types.

**Comment by lrzpellegrini** [comment]
> Ok, so I took a look at the PR. First of all, I believe that having proper types would help users and reduce the complexity a little bit. However, in order to make them useful, we have to make some design changes. Right now, some design choices result in very complex types. IMO the worse culprit are the circular dependencies that we have between experiences->streams->benchmark.

I agree. Currently, we have a "fully connected" dependency,
- benchmarks need to define the type of the stream and the experience
- streams need to define the type of the benchmark and the experience
- experiences need to define the type of the benchmark and the stream

We need to break this circularity, especially considering that the Python generics system does not support this kind of thing.

> I think we can remove them (i.e. not have `experience.origin_stream` and `stream.origin_benchmark`). This would require passing the benchmark object to the evaluator, which I think is a reasonable thing to do. The strategy should never touch the stream or benchmark object directly so it would be a benefit in this sense because it would remove the possibility of "cheating".

I think that those fields are useful in general for logging/debugging/etc. We can obfuscate them in the experiences by using `train()` and `eval()`. One thing that I did not describe in the PR is that **the experience mode now works as intended**!

> @lrzpellegrini what do you think? We don't have to do the change in this PR (actually, I would prefer a lot if we didn't, it's already huge), but we should make a choice.

For the typing circularity, I think it has to be fixed in this PR. As it currently is, the VSCode type checker (Pylance) is not able to properly understand what is going on and many things are not hinted as expected.

I'll try to:
- Remove the protocols. We currently have a parallel hierarchy of protocols that has little use.
- Break the circular generics dependency

I think that, by solving those 2 elements, the typing system will simplify a lot.

**Comment by AntonioCarta** [comment]
> I think that those fields are useful in general for logging/debugging/etc

They are, but:
- there are other ways to get the benchmark (i.e. you can always pass it around explicitly)
- they create implementation issues in several places (at least typing and initialization/creation/modification of benchmarks)

For logging, we can always pass the benchmark to the evaluator. Debugging may become a bit harder in some cases but I have to say that I don't see a good debugging usecase where it's helpful to go exp->stream->benchmark. You already have the experience ID in the experience which can be used to get to the same information if you have the benchmark (you can pass it around explicitly if needed).

Notice that those attributes are also creating other small implementation issues. For example, many times we create the experience before knowing its stream/benchmark, which means we have to patch the exp/stream after creation.

I think we reached the point where the disadvantages are more than the advantages of having them readily available.

> Remove the protocols. We currently have a parallel hierarchy of protocols that has little use.

I agree, this is another point that adds to the confusion.

**Comment by lrzpellegrini** [comment]
I think that, for the moment, the most balanced solution is to keep those fields in place so that we don't break things. At the same time, we can start by removing their type hint (and thus the need for their generic type). In this way, we can break the circular dependency.

If this is ok with you, I'll start adapting the PR today.

**Comment by AntonioCarta** [comment]
> I think that, for the moment, the most balanced solution is to keep those fields in place so that we don't break things. At the same time, we can start by removing their type hint (and thus the need for their generic type). In this way, we can break the circular dependency.

Ok, let's proceed this way. In this case, I would focus on providing a *minimal* set of types that can be really helpful for the users (e.g. avoiding too many generics and multiple ABC/protocols that are almost equivalent).

**Comment by lrzpellegrini** [comment]
Ok, generics are now much more simple! I'll proceed by fixing types in other modules. I just hope that merge conflicts will not be impossible to manage.

**Comment by lrzpellegrini** [comment]
@AntonioCarta we have probably just hit a corner case in the unit tests. Can you please check this :sweat_smile:?

Also, the PR is almost completed! I'll also check the logging and metrics packages but this will not take much time. I found a solution that allows us to have typing in the mixin classes (SupervisedProblem, BatchObservation, etc.). What do you think of that?

**Comment by AntonioCarta** [comment]
Hi Lorenzo, thanks for all the work.
I checked the PR. It's huge and there is a mix of refactoring and new features so I may have missed something.
- I would add an `@experimental` tag to the new features ([here](https://github.com/ContinualAI/avalanche/blob/master/avalanche/_annotations.py)). I'm not convinced by some choices, like the collate, that are a bit too complex
- I still think that if we want to use the types we should rethink some parts of the API to make types easier to understand (like removing benchmark and stream attributes)
- I'm going to open a separate issue for the broken test. I don't know why it's failing but I'm quite confident it's unrelated to your PR

Overall, I think it's a big improvement in terms of documentation. With your PR and the sphinx fix we should finally be able to document all the attributes in the strategy and benchmarks.

**Comment by lrzpellegrini** [comment]
I think we can postpone the collate thing. They will be needed to integrate the distributed training support, but for this PR I can remove it.

**Comment by AntonioCarta** [comment]
> I think we can postpone the collate thing. They will be needed to integrate the distributed training support, but for this PR I can remove it.

Whatever is more conveniente for you. Right now it's an internal interface so it shouldn't affect too many things. However, we need to discuss how we want to proceed more in detail before starting to use the new concat.

**Comment by lrzpellegrini** [comment]
The PR is ready for a final review. In the last commits, I fixed the types for the Metrics and added a minor fix for DownloadableDataset.

I also removed the redundant variables used to store the reference metric in the subclasses of GenericPluginMetric.

**Comment by AntonioCarta** [comment]
Thanks for all the work, everything looks ok.

Ideally, I would like to enforce types (e.g. having mypy in the CI) but right now I think they are a bit too complex to use and they may discourage new contributors. We should try to simplify them in future work.
